consensus/system_contract/system_contract.go

```diff
@@ -0,0 +1,102 @@
+package system_contract
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/scroll-tech/go-ethereum/common"
+	"github.com/scroll-tech/go-ethereum/log"
+	"github.com/scroll-tech/go-ethereum/params"
+	"github.com/scroll-tech/go-ethereum/rollup/sync_service"
+)
+
+const (
+	defaultSyncInterval = 10 * time.Second
+)
+
+// SystemContract
+type SystemContract struct {
+	config *params.SystemContractConfig // Consensus engine configuration parameters
+	client sync_service.EthClient
+
+	signerAddressL1 common.Address // Address of the signer stored in L1 System Contract
+
+	signer common.Address // Ethereum address of the signing key
+	signFn SignerFn       // Signer function to authorize hashes with
+	lock   sync.RWMutex   // Protects the signer and proposals fields
+
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+// New creates a SystemContract consensus engine with the initial
+// signers set to the ones provided by the user.
+func New(ctx context.Context, config *params.SystemContractConfig, client sync_service.EthClient) *SystemContract {
+	ctx, cancel := context.WithCancel(ctx)
+	address, err := client.StorageAt(ctx, config.SystemContractAddress, config.SystemContractSlot, nil)
+	if err != nil {
+		log.Error("failed to get signer address from L1 System Contract", "err", err)
+	}
+	systemContract := &SystemContract{
+		config:          config,
+		client:          client,
+		signerAddressL1: common.BytesToAddress(address),
+
+		ctx:    ctx,
+		cancel: cancel,
+	}
+	systemContract.Start()
+	return systemContract
+}
+
+// Authorize injects a private key into the consensus engine to mint new blocks
+// with.
+func (s *SystemContract) Authorize(signer common.Address, signFn SignerFn) {
+	log.Info("Authorizing system contract", "signer", signer.Hex())
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	s.signer = signer
+	s.signFn = signFn
+}
+
+func (s *SystemContract) Start() {
+	log.Info("starting SystemContract")
+	go func() {
+		syncTicker := time.NewTicker(defaultSyncInterval)
+		defer syncTicker.Stop()
+		for {
+			select {
+			case <-s.ctx.Done():
+				return
+			default:
+			}
+			select {
+			case <-s.ctx.Done():
+				return
+			case <-syncTicker.C:
+				address, err := s.client.StorageAt(s.ctx, s.config.SystemContractAddress, s.config.SystemContractSlot, nil)
```

Comment:
Would this address be different if the node is syncing at a different time and the signer in L1 has been changed? Would block verification fail during syncing historical blocks in the above case?

Reply:
The address of the contract should not be different AFAIK. The Signer may be different but that is ok. Verification will faill for new blocks but quickly after it will stop failing.

Reply:
oh, here I mean if the signer's address is updated. and the node is syncing historical blocks, would the signature verification fail due to the signer address mismatch (old signer address != new signer address)?

Reply:
Only new blocks are verified against the signature (messages sent via sendNewBlocks), so that should be no issue. This is done with the new internal field of the Header struct .IsNewBlock